File: DataStructureAndAlgorithm/string/kmp.java
/*

kmp

*/

public class kmp {

    /**
     * @param s 待匹配字符串
     * @return next数组
     */
    private static int[] get_next(final String s) {
        int[] next = new int[s.length()];
        next[0] = -1;
        for (int i = 1; i < next.length; i++) {
            int idx = next[i - 1];
            while (idx >= 0 && s.charAt(idx) != s.charAt(i - 1))
                idx = next[idx];
            next[i] = idx + 1;
        }
        return next;
    }

    /**
     * @param str 原串
     * @param s   待匹配字符串
     * @return s在str中的开始索引，倘若不存在则返回-1
     */
    private static int Kmp(final String str, final String s) {
        int[] next = get_next(s);
        int stri = 0, si = 0;
        while (stri < str.length() && si < s.length())
            if (str.charAt(stri) == s.charAt(si) || (si = next[si]) == -1) {
                stri++;
                si++;
            }
        return si >= s.length() ? stri - s.length() : -1;
    }

    public static void main(String[] args) {
        String str = "googloogle", s = "google";
        System.out.println(Kmp(str, s));

        str = "asdfgs";
        s = "dfg";
        System.out.println(Kmp(str, s));
    }
}

File: DataStructureAndAlgorithm/string/kmp.cpp
/*

kmp

*/

#include "iostream"
#include "string"
#include "vector"

// 获得next数组
std::vector<int> get_next(const std::string &s)
{
    std::vector<int> next(s.size(), -1);
    for (int i = 1; i < next.size(); i++)
    {
        int idx = next[i - 1];
        while (idx >= 0 && s[idx] != s[i - 1])
            idx = next[idx];
        next[i] = idx + 1;
    }
    return next;
}

// kmp
int kmp(const std::string &str, const std::string &s)
{
    auto &&next = get_next(s);
    int stri = 0, si = 0;
    while (stri < str.size() && si < s.size())
        if (str[stri] == s[si] || (si = next[si]) == -1)
            stri++, si++;
    return si >= s.size() ? stri - s.size() : -1;
}

int main(int argc, char *argv[])
{
    std::string str = "googloogle", s = "google";
    std::cout << kmp(str, s) << '\n';

    str = "asdasadffigk", s = "das";
    std::cout << kmp(str, s) << '\n';

    return 0;
}